### ISSUE TITLE
File deletion when editing a space over quota

### ISSUE THREAD

**Issue by dddns** [proftpd/proftpd#1185]
Howdy,

I use filezilla as ftp client.
I just hit something odd.
I use SQL quotas, and i was editing a file in a space that was over quota, and when i was about to send the edited file back to the FTP space, it just deleted the file along with the over quota warnings/information.
Is this known or happened in the past?
I'm using proftpd1.3.6c and last filezilla version.

**Comment by Castaglia** [comment]
This is the expected behavior; see the [`mod_quotatab` usage docs](http://www.proftpd.org/docs/contrib/mod_quotatab.html#Usage), particularly:
> A "hard" limit type means that a user's tally will never be allowed to exceed the limit; any file uploaded that exceeds the user's quota limit will be deleted, and the tally will not be incremented. A "soft" limit type, on the other hand, is a little less strict, for it will allow a user to exceed the quota limit by one file only. For example, assume a user's tally is close their limit. That user uploads a large file, one that pushes the tally up over the quota limit. If their limit type is "hard", the large file will be deleted once uploaded, and a message returned to the client informing them of this. If the user's limit type is "soft", the large file will be allowed, but any other uploads are denied (until the user deletes some files, and brings their tally back down below the quota limit).

**Comment by dddns** [comment]
Hi TJ,

OK understood the explanation, but, in my opinion i think this should have a redefine.

Scenario 1 (actual):
- the user is overquota, and tried to upload a "new" file, the file will be deleted.
- ok fine, i agree 100% that this should be the behavior or the user will be able to overquota even more.

Scenario 2 (what should happen, in my opinion):
- the user tried to edit a existing file, and when he tried to upload the file back, the FTP deletes the file.
- in this scenario, where the file already exists, the user SHOULD NOT be able to upload the edited file, but, proftpd should not delete the existing file.

Why i say this? I spent 2 hours on this day, and about 3 times in the last weeks recovering data for users, where they tried to edit their config.php of their CMS (wp or other) because the file was deleted.
Would it be possible to change the behavior to NOT delete the file, if it exists already?

**Comment by Castaglia** [comment]
Well, consider the case where a user keeps appending an already existing file, even though they are over their quota.  Your requested change would allow such users to continue exceeding their quota.

**Comment by dddns** [comment]
It's not possible to see, if the user is overquota, just deny append on existing files, and avoid the delete?

**Comment by Castaglia** [comment]
The `mod_quotatab` already _does_ deny the FTP `APPE` command, if the user is over the quota.

For your quota entries, do you have them configured for "hard" or "soft" limits?

**Comment by dddns** [comment]
it is with hard:

QuotaLimitTable                sql:/get-quota-limit

SQLNamedQuery get-quota-limit SELECT "name, quota_type, per_session, limit_type, bytes_in_avail, \
bytes_out_avail, bytes_xfer_avail, files_in_avail, files_out_avail, files_xfer_avail FROM quotalimits \
WHERE name = '%{0}' AND quota_type = '%{1}'"

![image](https://user-images.githubusercontent.com/13902147/110222350-816f1880-7ec9-11eb-89e1-1c3804039c1b.png)

**Comment by Castaglia** [comment]
You might consider changing that to use the "soft" limit; this tells `mod_quotatab` to be a little more lenient.

**Comment by dddns** [comment]
TJ,
Did the test with the soft type, and:

- yes it allows to edit and upload it again.
- BUT, if i upload 1.15GB file, it allows it aswell, and i'm already overquota, 2.3Gb over 1GB (files sent via HTTP).

TJ, there's really no way to change the "hard" type behavior, to just don't delete the file, if someone tries to upload a file with the same name and it already exists on the destination? System could just deny/refuse it, instead of deleting the file.
Or even a parameter for this behavior,

**Comment by Castaglia** [comment]
`mod_quotatab` only tracks data sent via FTP; it doesn't know anything about files uploaded via HTTP or anything else.  And it is not a filesystem quota system; you'd do better to use a filesystem-based quota, if that's what you actually need.

For a "hard" limit, if the upload exceeds the quota in mid-transfer, `mod_quotatab` breaks the upload.  If that file was not deleted, that would leave a partial file -- potentially corrupted -- on disk, which is arguably worse than no file at all.  That's why I consciously chose to have `mod_quotatab` force that "hard" limit this way.

**Comment by dddns** [comment]
TJ,

I understand your way of doing it.
I'm not going to filesystem-quota, all my systems use *sql in proftpd.

My problem is that, when a user does that, since it deletes the file, they are always asking me for help for that, that could be avoid if the system did not delete the file in this situation.

No way to make mod_quotatab know that an APPE if being done in a over quota session, and if yes, do nothing if the file exists?

**Comment by Castaglia** [comment]
This is where having the exact logs, of the FTP commands used by the client in question, might help.  Not all FTP clients use the `APPE` command; some clients implement such things far differently than you might expect.

**Comment by dddns** [comment]
TJ,

Sorry for the delay, was impossible for me to finish this before.
So what i did was:
- edit a file in an overquota space
- tried to put the file back, and proftpd deleted it due to overquota

[proftpd_log_overquota.txt](https://github.com/proftpd/proftpd/files/6254982/proftpd_log_overquota.txt)

**Comment by dddns** [comment]
Hi TJ,

Would it be possible to do some behavior change on this? please?

**Comment by dddns** [comment]
Hi TJ,
Sorry to re-ask, could you tell something about this possible feature?

**Comment by mdell-seradex** [comment]
From the log, it seems that the client is executing a STOR command to overwrite the file instead of appending to it.  According to the documentation, this is only allowed when `AllowOverwrite` is enabled. [proftpd.org docs - AllowOverwrite](http://www.proftpd.org/docs/modules/mod_xfer.html#AllowOverwrite)  My guess is that this causes the server to delete the file first.  I do not know if there is anything that allows the server to know the file size prior to the upload, but the client could lie about the file size, if that ability exists.
I wonder if enabling something such as `HiddenStores` [proftpd.org docs - HiddenStores](http://www.proftpd.org/docs/modules/mod_xfer.html#HiddenStores) could help with this, assuming that when enabled the original file is not deleted until the system attempts to rename the replacement, however, it is probably unlikely that this is what occurs.

**Comment by dddns** [comment]
@mdell-seradex don't disagree, but it should not delete the file.
If in a session where it's already over-quota, a STORE command should be denied and never delete the file.
or
In a over quota session, even if the STORE is not denied, should not delete the file

**Comment by mdell-seradex** [comment]
I can see the issue being that if the user is not over their hard quota prior to the STOR command, but the updated file pushes them over quota, that would be a conflict that the quota system could not anticipate to prevent the deletion of the original file.  Allowing the quota to work with the `HiddenStores` directive seems to me would be the better way to go, although I would think that proftpd would need to be configured to delete failed uploads to ensure that they would not cause them to expand the usage counted towards your quota.  I would also think that it would need to account for the fact that file X is being replace when calculating the usage to determine if the user is over their quota.  That also leads to another complication which is if the user is uploading multiple files simultaneously over separate connections.  Perhaps the user account is a shared one.  All of those things complicate determining a proper solution that cannot be configured to work incorrectly.
Maybe a better solution would be to have a quota child directive that behaves like `HiddenStores`, causes failed uploads to be deleted, and only allows a single connection per user at a time to prevent those complications.  It would have check the user's total usage, substituting the "hidden store" file for the one it is replacing in that calculation, and if the total usage does not go over the user's quota, only then would the original file be deleted and the "hidden store" file be renamed.  If it does go over the hard quota, then the "hidden store" file would be deleted and the quota error would occur.
Note that this would probably not be a quick and simple thing to achieve.  I also do not know of the validity of some of my ideas since I have only poked around a little inside proftpd's source code, but I have helped to identify and correct a few issues.

**Comment by Castaglia** [comment]
@dddns What's the point of having quotas, if the quotas cannot be enforced by deleting the over-quota files.

Also, as previously noted, the "hard" or "soft" nature of your configured quota limits determines the `mod_quotatab` behavior: a "hard" limit tells `mod_quotatab` to be strict, and to delete the files.  If you don't want files deleted, use a "soft" limit, not a "hard" limit.

**Comment by mdell-seradex** [comment]
My understanding was that the unmodified file did not put them over the quota, but the modified version did.
Because the file was being replaced, the file was lost instead of keeping the original version because of the order of operations and how the usage total is calculated, even though essentially the quota was preventing the replacement.  (delete original, try to save replacement, quota reach so block replacement)
I do believe that overall, @dddns wants it to behave like a hard quota, not allowing the final result to go over the limit.  Because this could occur with several files, allowing the limit to be exceeded by one file does not seem like it would work.

@dddns would need to confirm this.
However as I said, I don't think a solution would be simple.

**Comment by Castaglia** [comment]
In the current `mod_quotatab` implementation, _all_ write FTP commands (`APPE`, `STOR`, `STOU`) are denied if the session is already over quota, for both "soft" and "hard" limits.

So the discussion, then, is precisely how to handle the case where a client is doing a write that causes the session to go over quota.  Should the write be allowed to succeed (thus exceeding the quota), and all future writes blocked?  This is what the "soft" limit does.  Should the write be broken _at the point of going over the quota_?  This is what the "hard" limit does.  _In addition_, the "hard" limit implementation will delete the file being uploaded, to attempt to get back under the quota.  It is the last behavior which is the contentious part in this ticket, I think.

As I mentioned before, in this case of a "hard" limit, `mod_quotatab` _could_ let the write succeed -- in which case, we have allow the session to go over quota, and all future writes are immediately blocked.  But this is what the "soft" limit already does.  Or we could fail the write -- but leave the file on-disk as it is, and not delete it.  (This is what is being requested.)  However, in that case, the modified file might be incomplete, broken -- consider a partial HTML, XML, JSON file, script, _etc_.  Such partial files would cause unexpected errors/questions from users as well.  @dddns this partial-file scenario is exactly the behavior that you are requesting.

**Comment by mdell-seradex** [comment]
Something like the buffer/spool idea I think is what I think he was getting at, but perhaps instead doing of that with its potential issues you have mentioned, it could be done by renaming or archiving the original file that can be restored after if the quota fails the upload.  This way, the upload would not have to be "replayed" and there would be no concern of IO errors occurring in that case.  Yes an IO error could occur when renaming or archiving the file and then putting it back, but I think as long as it is on the same partition (I think it is still called that in Linux), the danger would be minimal.  If the user has access to the archive folder then they could restore it manually if the system was unable to for some reason.
The other way, which may be simpler, would be to have the new file be stored using a temporary name, and it would be given the correct name only after the quota validations succeed.  I thought I saw something about temporary file names for uploads in a proftpd module somewhere.  This way there would be no replay required.  Also, the user would not be given the final "ok" that the upload was accepted until after the validations succeed.

I see now that you included this in the "duplicate of the target file", and yes, of course if the file consumes the remaining disk space, that would be a different failure, perhaps "quota - out of disk space", if custom errors can be provided (I suspect they can).  I think that would likely be an acceptable issue because if they don't have much free space, that is likely a problem all on its own.  Simply adding another user would give them disk space problems.

No matter how it is decided to be done, if it even is, I still do not think this would be a simple change.

@dddns really needs to comment on this!

**Comment by Castaglia** [comment]
Yes, for uploading _new_ files to a hidden file first, the [`HiddenStores`](http://www.proftpd.org/docs/modules/mod_xfer.html#HiddenStores) directive would be used.  _Note_ that `HiddenStores` are _not_ usable when the FTP `APPE` command is used, since `HiddenStores` does not make a copy of existing files.  In fact, when `HiddenStores` is in effect, the `APPE` FTP command will be denied.

**Comment by mdell-seradex** [comment]
All that said, if @dddns is using a custom FTP client, then perhaps it would be better if the client backed up the files that are being uploaded first, either to an ftp folder or a local one.  That would eliminate the loss of files due to this and broken dependencies.
That said, is there a way to determine if the client is already over quota, because if they are, either because of soft limits or because if the quotas were changed and it cannot be detected, then it would not be possible to recover from a quota failure.

Note: The `HiddenStores` ability sounds great and that it would do what the user desires, assuming the original file size is excluded from the quota calculation.  Even if it is not, I believe it would be a far better place to be in then where @dddns is now.  🙂

**Comment by Castaglia** [comment]
> Is there a way to determine if the client is already over quota?

Yes; see the custom [`SITE QUOTA`](http://www.proftpd.org/docs/contrib/mod_quotatab.html#SITE_QUOTA) command.  Using this usually requires a custom FTP client.

**Comment by mdell-seradex** [comment]
It sounds like this behaviour would be an excellent thing to point out in the documentation, assuming it currently isn't.
If it is then we have missed something very important. 🙁 
🙂

**Comment by Castaglia** [comment]
I was thinking the same thing; see https://github.com/proftpd/proftpd/pull/1358

**Comment by dddns** [comment]
@Castaglia, @mdell-seradex  i was reading this now, was kinda off latelly.
Understood the explanation and makes sense.
I'll look into hiddenstore's and soft limits, and will configure it with the way i want to achieve.

Thank you both for this hard discussion on this matter.

**Comment by Castaglia** [comment]
@dddns is there any update?  Were you able to achieve the behavior you need?  I'd like to close this ticket, when possible.  Thanks!
